B-37 forgot password

shaguahill opened the pull request:
1.Created ResetPasswordServlet
2.Added reset-password page and link to this page from login page
3.Added language properties to reset-password page
4.Changed scripts V1_12, V1_13 for DB witch didn`t understand h2 db
5. Added cache for messages about password and mapper for it
6. Added table to db PasswordEmailTemplate
7. added enum MessageType for email types and entity PasswordEmailTemplate
8. Added method resetUserPassword and moved method change password to Security folder
9. Added method sendNewPassword to MailService
10. Added languages properties for change-password page
11. Added tests for EmailTemplateCachedDao for ResetPassword

Bug fix [def-19]:
1. Created method changePassword in UserDao
2. Changed logic for method changeUserPassword and ChangePasswordServlet

shaguahill committed:
[b-37] Forgot password:
Created ResetPasswordServlet
Added reset-password page and link to this page from login page
Added language properties to reset-password page
Changed scripts V1_12, V1_13 for DB witch didn`t understand h2 db

Bug fix [def-19]:
Created method changePassword to UserDao
changed logic for method changeUserPassword and ChangePasswordServlet

shaguahill committed:
Merge branch 'develop' into b-37_forgot_password

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

shaguahill committed:
b-37_forgot_password
1. Added cache for messages about password and mapper for it
2. Added table to db PasswordEmailTemplate
3. added enum MessageType for email types and entity PasswordEmailTemplate
4. Added method resetUserPassword and moved method change password to Security folder
5. Added method sendNewPassword to MailService
5. fixed design reset-password page

shaguahill committed:
Merge branch 'develop' into b-37_forgot_password

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

shaguahill committed:
b-37_forgot_password:
1. Added languages properties for change-password page
2. Added tests for EmailTemplateCachedDao for ResetPassword

Rooman commented on src/main/java/com/photostudio/dao/jdbc/JdbcEmailTemplateCachedDao.java:
```diff
@@ -55,4 +83,18 @@ void load(DataSource dataSource) {
+        try (Connection connection = dataSource.getConnection();
+             PreparedStatement preparedStatement = connection.prepareStatement(GET_ALL_PASSWORD_EMAIL_TEMPLATES);
+             ResultSet resultSet = preparedStatement.executeQuery()) {
+            passwordEmailTemplateCache = new ArrayList<>();
```
1. Плохая идея ассайнить пустую коллекцию. Может быть обращение к пустой из другого потока
2. volatile

shaguahill replied:
Fixed. О volatile перечитал.

Rooman commented on src/main/java/com/photostudio/dao/jdbc/JdbcEmailTemplateCachedDao.java:
```diff
@@ -39,6 +53,20 @@ public EmailTemplate getByLangAndStatus(int langId, OrderStatus orderStatus) {
+                .filter(passwordEmailTemplate -> passwordEmailTemplate.getMessageType() == messageType)
+                .filter(passwordEmailTemplate -> passwordEmailTemplate.getLangId() == langId)
+                .findFirst();
+        if (optionalPasswordEmailTemplate.isPresent()) {
```
if(validate) {
   throw exception
}

normal logic

Yuliia-Sh replied:
тут можна один filter использовать

shaguahill replied:
Fiexed.
О использовании filter - в курсе. Использовал 2 для читабельности кода.

Rooman commented on src/main/java/com/photostudio/entity/email/MessageType.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public int getMessageId() {
+        return messageId;
```
lombok/

shaguahill replied:
Добавил

Rooman commented on src/main/java/com/photostudio/entity/email/MessageType.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public static MessageType getByMessageType(String messageTypeName) {
+
```
bez perenosov strok nikyda

shaguahill replied:
Убрал лишний перенос

Rooman commented on src/main/java/com/photostudio/entity/email/MessageType.java:
```diff
@@ -0,0 +1,39 @@
+        return Arrays.stream(MessageType.values())
+                .filter(messageType -> messageType.nameMessageType.equalsIgnoreCase(messageTypeName))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("MessageTpe name is not correct!"));
```
a kakoy imenno name ne correct/

shaguahill replied:
Добавил названия в ошибки

Rooman commented on src/main/java/com/photostudio/security/impl/DefaultSecurityService.java:
```diff
@@ -107,6 +110,22 @@ public boolean isOldPassword(String inputOldPassword, User user) {
+
+    @Override
+    public void resetUserPassword(User user) {
+        log.info("Started service reset password for user with id " + user.getId());
```
"+"

shaguahill replied:
Косяк. Устранил

Rooman commented on src/main/java/com/photostudio/service/impl/DefaultMailService.java:
```diff
@@ -38,6 +40,15 @@ public void sendOnChangeStatus(User userChanged, int orderId, OrderStatus orderS
+        PasswordEmailTemplate passwordEmailTemplate = emailTemplateDao
+                .getPasswordEmailTemplateByLangIdAndMessageType(user.getLangId(), MessageType.RESET_PASSWORD);
+        mailSender.send(passwordEmailTemplate.getSubject(), passwordEmailTemplate.getBody()
+                .replace("<password>", password), user.getEmail());
```
:+1:

Rooman commented on src/main/java/com/photostudio/web/servlet/security/ResetPasswordServlet.java:
```diff
@@ -0,0 +1,54 @@
+
+        String email = request.getParameter("email");
+
+        User user = userService.getUserByEmail(email);
```
1 method v servlete = 1 vuzov metoda service. 
servlet ne yavlyaetya dispatcherom. logica dolgna but v service

shaguahill replied:
Изменил логику

Rooman commented on src/main/webapp/assets/css/reset-password.css:
```diff
@@ -0,0 +1,75 @@
+    font-weight: bold;
+    color: black !important;
+    background-color: #FCC5D9 !important;
+    border-color: #FCC5D9 !important;
```
ваши стили очень важны для нас...

shaguahill replied:
Хотя мой css файл после bootstrap, при active bootstrap имеет выше приоритет. По этому !important (P.S. В курсе что решение не очень)

Rooman replied:
нужно было поставить более сложный селектор 
ul.items li a.item-link  {}

Yuliia-Sh commented on src/main/java/com/photostudio/security/impl/DefaultSecurityService.java:
```diff
@@ -24,18 +25,20 @@
-        this(ServiceLocator.getService(UserService.class), ServiceLocator.getService(MailSender.class));
+        this(ServiceLocator.getService(UserService.class), ServiceLocator.getService(MailSender.class), ServiceLocator.getService(MailService.class));
     }
 
```
MailService содержит MailSender, тут точно нужно и то и другое?

shaguahill replied:
Он нужен для метода register для добавления нового юзера. Когда мы перенесем отправку сообщение для этого случая в MailService - MailSender уберется.

shaguahill committed:
b-37_forgot_password (review fix):
Added lombok to MessageType
Removed "+" from logs
Fixed problem with passwordEmailTemplateCache
Added error to order page when email is not correct
format order page

Miari commented on src/main/webapp/WEB-INF/templates/change-password_de.properties.properties:
```diff
@@ -0,0 +1,9 @@
+page.title=Foto Studio
+change.password.title=Ändere das Passwort
+change.password.label.old=Altes Passwort
+change.password.label.new=Neue Passwort
```
Neues Passwort

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_de.properties.properties:
```diff
@@ -0,0 +1,9 @@
+page.title=Foto Studio
+change.password.title=Ändere das Passwort
```
Passwort ändern

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password.properties:
```diff
@@ -0,0 +1,9 @@
+page.title=Photo Studio
```
Photostudio

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_de.properties.properties:
```diff
@@ -0,0 +1,9 @@
+page.title=Foto Studio
```
Fotostudio

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_de.properties.properties:
```diff
@@ -0,0 +1,9 @@
+change.password.title=Ändere das Passwort
+change.password.label.old=Altes Passwort
+change.password.label.new=Neue Passwort
+change.password.label.repeat.new=Wiederhole das neue Passwort
```
Neues Passwort wiederholen

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_de.properties.properties:
```diff
@@ -0,0 +1,9 @@
+change.password.label.old=Altes Passwort
+change.password.label.new=Neue Passwort
+change.password.label.repeat.new=Wiederhole das neue Passwort
+change.password.error.incorrect.old=Das Passwort ist inkorrekt
```
Das Passwort ist falsch

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password.properties:
```diff
@@ -0,0 +1,9 @@
+change.password.label.new=New password
+change.password.label.repeat.new=Repeat new password
+change.password.error.incorrect.old=The password is incorrect
+change.password.error.not.match=Password does not match
```
The password does not match

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_de.properties.properties:
```diff
@@ -0,0 +1,9 @@
+change.password.label.new=Neue Passwort
+change.password.label.repeat.new=Wiederhole das neue Passwort
+change.password.error.incorrect.old=Das Passwort ist inkorrekt
+change.password.error.not.match=Passwort stimmt nicht überein
```
Das Passwort stimmt nicht überein

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_en_US.properties.properties:
```diff
@@ -0,0 +1,9 @@
+page.title=Photo Studio
```
Photostudio

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_en_US.properties.properties:
```diff
@@ -0,0 +1,9 @@
+change.password.label.new=New password
+change.password.label.repeat.new=Repeat new password
+change.password.error.incorrect.old=The password is incorrect
+change.password.error.not.match=Password does not match
```
The password does not match

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_ru.properties.properties:
```diff
@@ -0,0 +1,9 @@
+change.password.label.new=Новый пароль
+change.password.label.repeat.new=Повторить новый пароль
+change.password.error.incorrect.old=Пароль некорректный
+change.password.error.not.match=Паролии не совпадают
```
Пароли не совпадают

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_ru.properties.properties:
```diff
@@ -0,0 +1,9 @@
+page.title=Фото Студия
```
Фотостудия

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/change-password_ru.properties.properties:
```diff
@@ -0,0 +1,9 @@
+change.password.title=Изменить пароль
+change.password.label.old=Старый пароль
+change.password.label.new=Новый пароль
+change.password.label.repeat.new=Повторить новый пароль
```
Повторите новый пароль

shaguahill replied:
Fixed

Rooman replied:
@shaguahill ты каждый раз писал фиксел или копировал и вставлял? )

Miari commented on src/main/webapp/WEB-INF/templates/reset-password.properties:
```diff
@@ -0,0 +1,6 @@
+page.title=Photo Studio
```
Photostudio

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/reset-password_de.properties:
```diff
@@ -0,0 +1,6 @@
+reset.password.label=Geben Sie Ihre E-Mail-Adresse ein und wir senden Ihnen ein neues Passwort.
+reset.password.placeholder=Geben sie ihre E-Mailadresse ein
+reset.password.error=Benutzer wurde nicht gefunden
+reset.password.button=Neues Passwort senden
```
Fotostudio

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/reset-password_en_US.properties:
```diff
@@ -0,0 +1,6 @@
+page.title=Photo Studio
```
Photostudio

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/reset-password_ru.properties:
```diff
@@ -0,0 +1,6 @@
+page.title=Фото Студия
```
Фотостудия

shaguahill replied:
Fixed

Miari commented on src/main/webapp/WEB-INF/templates/reset-password_de.properties:
```diff
@@ -0,0 +1,6 @@
+page.title=Foto Studio
+reset.password.title=Passwort zurücksetzen
+reset.password.label=Geben Sie Ihre E-Mail-Adresse ein und wir senden Ihnen ein neues Passwort.
+reset.password.placeholder=Geben sie ihre E-Mailadresse ein
```
Geben Sie Ihre E-Mail-Adresse ein

shaguahill replied:
Fixed

Miari commented on src/test/java/com/photostudio/dao/jdbc/JdbcEmailTemplateCachedDaoTest.java:
```diff
@@ -43,6 +44,23 @@ void testLoad() {
+    @Test
+    void testGetPasswordEmailTemplateByLangIdAndMessageType() {
+        //prepare
+        String expectedSubject = "Miari Fotografie Reset password";
```
А Subject на немецком мы проверяем где-нибудь?

shaguahill replied:
Subject и body берутся с DB. Файл с накаткой V1_15__add_password_email_templates.sql

shaguahill committed:
b-37_forgot_password (review fix):
Fixed languages

Rooman commented:
красавчик :+1:
